smdebug/tensorflow/utils.py

```diff
@@ -261,6 +297,53 @@ def get_keras_layer_inputs(layer):
     return input_tensors
 
 
+class LayerWithHooks(tf.keras.layers.Layer):
+    def __init__(
+        self,
+        layer: tf.keras.layers.Layer,
+        hooks: List[Callable[[tf.Tensor, tf.Tensor], Optional[tf.Tensor]]] = None,
+    ):
+        super().__init__()
+        self._layer = layer
+        self._hooks = hooks or []
+
+    def call(self, input: tf.Tensor) -> tf.Tensor:
+        output = self._layer(input)
+        for hook in self._hooks:
+            hook_result = hook(input, output)
+            if hook_result is not None:
+                output = hook_result
+        return output
+
+    def register_hook(self, hook: Callable[[tf.Tensor, tf.Tensor], Optional[tf.Tensor]]) -> None:
+        self._hooks.append(hook)
+
+
+class InputOutputSaver:
+    def __init__(self):
+        self.layer_input = None
+        self.layer_output = None
+
+    def __call__(self, callable_inputs, *args, **kwargs) -> None:
+        self.layer_input = kwargs["layer_input"]
+        self.layer_output = kwargs["layer_output"]
+
+
+def get_layer_call_fn(layer: tf.keras.layers.Layer) -> Callable[[tf.Tensor], tf.Tensor]:
+    old_call_fn = layer.call
+
+    def call(callable_inputs, *args, **kwargs) -> tf.Tensor:
```

Comment:
when I was implementing wrap tape, I faced this problem where the tape was getting recursively wrapped. (thats why theres a unwrap_tape). is there a chance something similar will happen here?

Reply:
This is prevented by saving state to `self.has_registered_model`